src/main/java/org/tron/core/db/Manager.java

```diff
@@ -965,17 +965,29 @@ public boolean processTransaction(final TransactionCapsule trxCap, Block block)
       throw new ValidateSignatureException("trans sig validate failed");
     }
 
-    /**  VM execute  **/
+    TransactionTrace trace = new TransactionTrace(trxCap);
+    trace.init();
+
 
     DepositImpl deposit = DepositImpl.createRoot(this);
     Runtime runtime;
 
-    runtime = new Runtime(trxCap.getInstance(), block, deposit,
+    runtime = new Runtime(trace, block, deposit,
         new ProgramInvokeFactoryImpl());
     consumeBandwidth(trxCap, runtime.getResult().getRet());
+
     runtime.init();
-    runtime.execute();
-    runtime.go();
+        
+    //exec
+    trace.exec(runtime);
+
+    //check SR's bill and ours.
+    if(block != null) {
+      trace.checkBill();
+    }
+
+    trace.finalize();
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Remove this call to finalize(). ([squid:ObjectFinalizeCheck](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AObjectFinalizeCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1192&issues=AWThGfwxOVxoJwmFD07H&open=AWThGfwxOVxoJwmFD07H)

[ISSUE_KEY]: AWThGfwxOVxoJwmFD07H